soc/arm/atmel_sam0/common/soc_gpio.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2016 Piotr Mienkowski
+ * Copyright (c) 2018 Google LLC.
+ * Copyright (c) 2021 Gerson Fernando Budke
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief Atmel SAM0 MCU family I/O Pin Controller (PORT)
+ */
+
+#include "soc_gpio.h"
+
+int soc_gpio_pinmux_set(PortGroup *pg, uint32_t pin, uint32_t func)
+{
+	bool is_odd = pin & 1;
+	int idx = pin / 2U;
+
+	/* Each pinmux register holds the config for two pins.  The
+	 * even numbered pin goes in the bits 0..3 and the odd
+	 * numbered pin in bits 4..7.
+	 */
+	if (is_odd) {
+		pg->PMUX[idx].bit.PMUXO = func;
+	} else {
+		pg->PMUX[idx].bit.PMUXE = func;
+	}
+	pg->PINCFG[pin].bit.PMUXEN = 1;
+
+	return 0;
+}
+
+void soc_gpio_configure(const struct soc_gpio_pin *pin)
+{
+	PortGroup *pg = pin->regs;
+	uint32_t flags = pin->flags;
+	uint32_t func =  (pin->flags & SOC_GPIO_FUNC_MASK) >> SOC_GPIO_FUNC_POS;
+
+	/* Reset or analog I/O: all digital disabled */
+	pg->PINCFG[pin->pinum].reg = 0;
+	pg->DIRCLR.reg = (1 << pin->pinum);
+	pg->OUTCLR.reg = (1 << pin->pinum);
+
+	if (flags & SOC_GPIO_PMUXEN_ENABLE) {
+		soc_gpio_pinmux_set(pg, pin->pinum, func);
+		return;
```

Comment:
Do we really want to exit? Shouldn't we process remaining flags?

Reply:
Hi @mnkp ,

I took in consideration below part from datasheet:

_21. PORT
21.1 Overview
The Port (PORT) controls the I/O pins of the micro controller. The I/O pins are organized in a series of groups, collectively
referred to as a port group, and each group can have up to 32 pins that can be configured and controlled individually or
as a group. Each pin may either be used for general-purpose I/O under direct application control or assigned to an
embedded device peripheral. When used for general-purpose I/O, each pin can be configured as input or output, with
highly configurable driver and pull settings._

By my understanding, if user selects a peripheral that will control the lines. I look at [spi_sam0_configure](https://github.com/zephyrproject-rtos/zephyr/blob/e62c3c533a37a4e7a413500a6375a350306e1e0f/drivers/spi/spi_sam0.c#L68-L109), for instance,  and the pin directions is made by SERCOM and not by PORT.

@stephanosio , do you confirm code or there are PORT pin configs that may be necessary configure?

Reply:
Good point. Typically the pin controller/multiplexer module when it connects a pin to a peripheral will let the peripheral set the pin direction but still control other pin properties like pull up / pull down, open drain configuration, etc. That doesn't seem to be the case for SAM0 PORT module. When PORT routes a pin signal to a peripheral it gives it full control of the pin configuration. E.g. SDA pin on I2C bus will be configured as open drain by the SERCOM and not PORT module.

But this means - taking into account that gpio_sam0.c driver is accessing PORT directly - that we cannot support any flags in pinmux mode and ultimately the `soc_gpio_configure` is not needed. We will need only `soc_gpio_pinmux_set`. Also all the related DT properties in `atmel,sam0-pinctrl.yaml` will have to be removed.

By the way, shouldn't we use `soc_port.c` for the file name and corresponding prefix for functions in this file?

Reply:
> But this means - taking into account that gpio_sam0.c driver is accessing PORT directly - that we cannot support any flags in pinmux mode and ultimately the soc_gpio_configure is not needed. We will need only soc_gpio_pinmux_set. Also all the related DT properties in atmel,sam0-pinctrl.yaml will have to be removed.

Questions:

1) Is there any possibility to use in future pinctrl to configure gpio pins? This structure allow exactly this. So, gpio API could inspect the pinctrl flags etc and we can have only one structure to define pins.

2) Could this help to manage Low Power Mode pins?

Example:
```
/* User define pin */
	pa19_port_gpio: pa19_port_gpio {
		atmel,pins = < &porta 0x13 0x0 >;
		output-low;
		/* pinmux-enable; This disable PORT and enable PINMUX */
	};
...
	leds {
		compatible = "gpio-leds";
		led0: led_0 {
-			gpios = <&porta 19 GPIO_ACTIVE_LOW>;
+			pinctrl-0 = <&pa19_port_gpio>;
			label = "Yellow LED";
		};
	};
```

> By the way, shouldn't we use soc_port.c for the file name and corresponding prefix for functions in this file?

Yes, make sense, I'll changed it.

Reply:
> Could this help to manage Low Power Mode pins?

Interesting. That's not so clear looking at Linux pinctrl bindings. But indeed, having ability to connect pin to GPIO (PORT module in case of SAM0) and configure it via pinctrl API will help handling low power modes. We should support it. Let's keep the current implementation. Thanks!

Reply:
One more thing. Shouldn't we add `else` branch to this `if` with
```
pg->PINCFG[pin->pinum].bit.PMUXEN = 0;
```

Reply:
tks @mnkp,
I fixed the variable initialization.